### ISSUE TITLE
Gradients from `jacrev` and `jacfwd` not matching Finite Differences

### ISSUE THREAD

**Issue by Microno95** [jax-ml/jax#29111]
### Description

Gradients computed via `jacrev` and `jacfwd` for the following function when evaluated at `[1.0, 0.0]` are not correct as tested by `jax.lax.check_grads`.

A minimal reproducer is as follows:

```python
import jax.test_util

def fn(x):
    return jnp.sum(jnp.abs(x))**4+jnp.exp(x[...,1])

x0 = jnp.array([1.0, 0.0])
jax.test_util.check_grads(fn, (x0, ), order=2)
```

It's unclear to me why this would specifically cause issues, what is the best way to detect when such a function fails?

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.5.3
jaxlib: 0.5.3
numpy:  2.2.6
python: 3.12.10 (main, Apr  9 2025, 00:30:25) [GCC 10.2.1 20210110]
device info: NVIDIA GeForce RTX 3070 Laptop GPU-1, 1 local devices"
process_count: 1
platform: uname_result(system='Linux', node='c0fd015d929e', release='6.6.87.1-microsoft-standard-WSL2', version='#1 SMP PREEMPT_DYNAMIC Mon Apr 21 17:08:54 UTC 2025', machine='x86_64')


$ nvidia-smi
Fri May 30 11:15:42 2025       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 575.51.03              Driver Version: 576.28         CUDA Version: 12.9     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA GeForce RTX 3070 ...    On  |   00000000:01:00.0  On |                  N/A |
| N/A   66C    P0             30W /   91W |    2833MiB /   8192MiB |      5%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
                                                                                         
+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI              PID   Type   Process name                        GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|    0   N/A  N/A              24      G   /Xwayland                             N/A      |
|    0   N/A  N/A              33      G   /Xwayland                             N/A      |
+-----------------------------------------------------------------------------------------+
```

**Comment by jakevdp** [comment]
`abs(x)` has a discontinuity in its gradient at `x = 0`, and so its second derivative is not well-defined at this point. Given this, it's not surprising that second-order gradients of functions including `abs` evaluated at `0.0` wouldn't match finite difference results.

**Comment by Microno95** [comment]
Ah, that's entirely my bad, you're absolutely right. I guess I did find it surprising that, given the hessian everywhere is 0, that at x=0, the gradient is equal to 1 although in hindsight it makes some sense for it to be positive given that the jacobian flips in sign at that point.
